compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BirVariableOptimizer.java

```diff
@@ -0,0 +1,238 @@
+        unusedVars.forEach(var -> birFunction.localVars.remove(var));
+    }
+
+    private void reuseVars(Map<BIRAbstractInstruction, Set<BIRNode.BIRVariableDcl>> liveOuts,
+                           List<BIRNode.BIRVariableDcl> unusedVars, BIRNode.BIRFunction birFunction) {
+        List<BIRAbstractInstruction> instructionList = getInstructionList(birFunction);
+        Map<BType, LinkedList<BIRNode.BIRVariableDcl>> freeVars = new HashMap<>();
+        for (int i = 0; i < instructionList.size(); i++) {
+            if (instructionList.get(i).getKind() == InstructionKind.XML_SEQ_STORE) {
+                continue;
+            }
+            tryToReuseFromFreeVars(liveOuts, freeVars, instructionList, i, unusedVars);
+            checkForFreeVars(freeVars, liveOuts, instructionList.get(i));
+        }
+    }
+
+    private void checkForFreeVars(Map<BType, LinkedList<BIRNode.BIRVariableDcl>> freeVars,
+                                  Map<BIRAbstractInstruction, Set<BIRNode.BIRVariableDcl>> liveOuts,
+                                  BIRAbstractInstruction instruction) {
+        for (BIROperand operand : instruction.getRhsOperands()) {
+            BType type = operand.variableDcl.type;
+            if (operand.variableDcl.kind == VarKind.TEMP && !liveOuts.get(instruction).contains(operand.variableDcl)) {
+                if (!freeVars.containsKey(type)) {
+                    freeVars.put(type, new LinkedList<>());
+                }
+                LinkedList<BIRNode.BIRVariableDcl> ls = freeVars.get(type);
+                ls.add(operand.variableDcl);
+            }
+        }
+    }
+
+    private void tryToReuseFromFreeVars(Map<BIRAbstractInstruction, Set<BIRNode.BIRVariableDcl>> liveOuts,
+                                        Map<BType, LinkedList<BIRNode.BIRVariableDcl>> freeVars,
+                                        List<BIRAbstractInstruction> instructionList, int idx,
+                                        List<BIRNode.BIRVariableDcl> unusedVars) {
+        BIRAbstractInstruction instruction = instructionList.get(idx);
+        if (instruction.lhsOp != null) {
+            BType type = instruction.lhsOp.variableDcl.type;
+            LinkedList<BIRNode.BIRVariableDcl> defLs = freeVars.get(type);
+            if (defLs != null && !defLs.isEmpty()) {
+                BIRNode.BIRVariableDcl newLhsOp = defLs.peek();
+                BIRNode.BIRVariableDcl oldLhsOp = (instruction).lhsOp.variableDcl;
+                if (oldLhsOp.kind == VarKind.TEMP && !checkVarUse(instruction, oldLhsOp)) {
+                    defLs.remove();
+                    unusedVars.add(oldLhsOp);
+                    replaceOldOp(instructionList, idx, oldLhsOp, newLhsOp, liveOuts);
+                    instruction.lhsOp.variableDcl = newLhsOp;
+                }
+            }
+
+        }
+    }
+
+    private void replaceOldOp(List<BIRAbstractInstruction> ls, int idx, BIRNode.BIRVariableDcl oldLhsOp,
```

Comment:
what is idx here?

Reply:
and we should use proper name for "ls" here.

Reply:
idx is the index, I have renamed if unclear